presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1465,29 +1468,38 @@ protected Scope visitSampledRelation(SampledRelation relation, Optional<Scope> s
                     accessControl,
                     sqlParser,
                     TypeProvider.empty(),
-                    ImmutableList.of(relation.getSamplePercentage()),
+                    ImmutableList.of(samplePercentage),
                     analysis.getParameters(),
                     WarningCollector.NOOP,
                     analysis.isDescribe())
                     .getExpressionTypes();
 
-            ExpressionInterpreter samplePercentageEval = expressionOptimizer(relation.getSamplePercentage(), metadata, session, expressionTypes);
+            Type samplePercentageType = expressionTypes.get(NodeRef.of(samplePercentage));
+            if (!typeCoercion.canCoerce(samplePercentageType, DOUBLE)) {
+                throw semanticException(TYPE_MISMATCH, samplePercentage, "Sample percentage should be a numeric expression");
+            }
+
+            ExpressionInterpreter samplePercentageEval = expressionOptimizer(samplePercentage, metadata, session, expressionTypes);
 
             Object samplePercentageObject = samplePercentageEval.optimize(symbol -> {
-                throw semanticException(EXPRESSION_NOT_CONSTANT, relation.getSamplePercentage(), "Sample percentage cannot contain column references");
+                throw semanticException(EXPRESSION_NOT_CONSTANT, samplePercentage, "Sample percentage cannot contain column references");
             });
 
-            if (!(samplePercentageObject instanceof Number)) {
-                throw semanticException(TYPE_MISMATCH, relation.getSamplePercentage(), "Sample percentage should evaluate to a numeric expression");
+            ResolvedFunction coercion = metadata.getCoercion(samplePercentageType, DOUBLE);
```

Comment:
This is only needed if `samplePercentageType != DOUBLE`

Reply:
Also, this could be done after checking whether the value is null, in which case, you don't need a separate variable for `samplePercentageObject`. An implicit cast should never produce a null from a non-null value.

Reply:
> An implicit cast should never produce a null from a non-null value.

I thought about this, but i didn't want to put this assumption in the code,
on the assumption that less assumptions the easier to read...

i will rearange if you find this more useful that way